202318033
Harshil shah

#• RDD Tasks

In [ ]:
!pip install pyspark

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 317.0/317.0 MB 3.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for pyspark: filename=pyspark-3.5.1-py2.py3-none-any.whl size=317488493 sha256=0f7d459d0f0e990cabe0ab5a4dbee6b911b5ceadc3f5db60a49ee5a3d968ff7c
  Stored in directory: /root/.cache/pip/wheels/80/1d/60/2c256ed38dddce2fdd93be545214a63e02fbd8d74fb0b7f3a6
Successfully built pyspark


##Task-1

-->Generate 100 random numbers in range 0 to 10 using numpy
randint function with the seed set to 10.

In [ ]:
import numpy as np

# Generate random numbers using numpy with seed set to 10
np.random.seed(10)
random_numbers = np.random.randint(0, 11, 100, dtype=int)
print(random_numbers)

[ 9  4  0  1  9  0  1 10  8  9  0 10  8  6  4  3  0  4  6  8 10  1  8  4
  1  3  6  5  3  9  6  9  1  9  4  2  6  7  8 10  8  9  2  0  6  7  8  1
  7  1  4 10  0  8  5  4  7  8  8  2  6  2  8  8  6  6  5 10  6  0  0  6
  9  1  8 10  9  1  2  8  9  9  5  0  2  7  3  0  4  2  0  3  3  1  2  5
  9  0 10  1]


-->Create a RDD using the paral-
lelize function using data generated in previous step.

In [ ]:
from pyspark import SparkContext

#SparkContext
sc = SparkContext("local", "RDD Task 1")

# Create RDD using parallelize function
# parallelize function is used to create an RDD
rdd = sc.parallelize(random_numbers)



-->Calculate the fre-quency of each number (0 - 10) using appropriate function of RDD.

In [ ]:
# Calculate the frequency of each number (0 - 10)
freq_count = rdd.countByValue()

# Print the frequency of each number
for num, count in sorted(freq_count.items()):
    print(f"Number {num}: {count} occurrences")

# Stop the SparkContext
sc.stop()


Number 0: 12 occurrences
Number 1: 11 occurrences
Number 2: 8 occurrences
Number 3: 6 occurrences
Number 4: 8 occurrences
Number 5: 5 occurrences
Number 6: 11 occurrences
Number 7: 5 occurrences
Number 8: 14 occurrences
Number 9: 12 occurrences
Number 10: 8 occurrences


##TASK-2:

In [ ]:
from pyspark import SparkContext

sc = SparkContext("local", "WordFrequencyApp")

#Purpose: Loads the text8 dataset into an RDD.
text_data = sc.textFile("/content/drive/MyDrive/DATASCIENCE/sem2/BD/pyspark/text8")


# Purpose: Splits each line into words and flattens the result.
words_rdd = text_data.flatMap(lambda line: line.split(" "))

# Purpose: Filters words that contain the letter 'a'.
filtered_words_rdd = words_rdd.filter(lambda word: 'a' in word)


# Purpose: Maps each word to a key-value pair, where the word is the key and 1 is the value.
word_count_rdd = filtered_words_rdd.map(lambda word: (word, 1))

# The reduceByKey transformation aggregates the counts of each word by adding up the values associated with each key (word).
word_frequencies = word_count_rdd.reduceByKey(lambda x, y: x + y)

# Print the results
for word, frequency in sorted(word_frequencies.collect()):
    print(f"Word '{word}': {frequency} occurrences")

# Stop the SparkContext (important to do at the end of your Spark application)
sc.stop()


Streaming output truncated to the last 5000 lines.
Word 'wachovia': 3 occurrences
Word 'wachowicz': 1 occurrences
Word 'wachowski': 11 occurrences
Word 'wachs': 2 occurrences
Word 'wachsmann': 2 occurrences
Word 'wacht': 3 occurrences
Word 'wachtelhund': 2 occurrences
Word 'wachtler': 1 occurrences
Word 'wachtmeister': 1 occurrences
Word 'wachuku': 2 occurrences
Word 'wachusett': 1 occurrences
Word 'wack': 3 occurrences
Word 'wacker': 1 occurrences
Word 'wackett': 1 occurrences
Word 'wackford': 1 occurrences
Word 'wackiest': 1 occurrences
Word 'wackiki': 1 occurrences
Word 'wackiness': 1 occurrences
Word 'wacking': 1 occurrences
Word 'wacko': 2 occurrences
Word 'wacky': 15 occurrences
Word 'wackyland': 6 occurrences
Word 'waclaw': 1 occurrences
Word 'waco': 40 occurrences
Word 'wacquant': 1 occurrences
Word 'wad': 10 occurrences
Word 'wada': 3 occurrences
Word 'wadada': 1 occurrences
Word 'wadadli': 1 occurrences
Word 'wadaiko': 1 occurrences
Word 'wadal': 1 occurrences
Word 'wadb': 6 

#• DataFrame Task:

In [ ]:
from pyspark.sql import SparkSession
from pyspark.sql.types import StructType, StructField, FloatType, StringType

spark = SparkSession.builder.master("local").appName('irisfile').getOrCreate()

# Load the iris JSON data into a DataFrames
df = spark.read.option("mode", "PERMISSIVE").json("/content/drive/MyDrive/DATASCIENCE/sem2/BD/pyspark/iris.json")
''' "PERMISSIVE" mode is used when encountering corrupt records (records that do not follow the expected schema). In this
mode, Spark tries to parse and load the data, and if it encounters corrupt records, it records them in a special column called _corrupt_record
and continues loading the rest of the data.'''


# Print the DataFrame schema
df.printSchema()

# Print a few lines of the raw JSON data
df.show(truncate=False)

# Define an explicit schema
custom_schema = StructType([
    StructField("petalLength", FloatType(), True),
    StructField("petalWidth", FloatType(), True),
    StructField("sepalLength", FloatType(), True),
    StructField("sepalWidth", FloatType(), True),
    StructField("species", StringType(), True)
])

# Load the iris JSON data into a DataFrame with the explicit schema
df_new = spark.read.schema(custom_schema).json("/content/drive/MyDrive/DATASCIENCE/sem2/BD/pyspark/iris.json")

# Print the DataFrame schema
df_new.printSchema()
df_new = df_new.dropna()

# Show the DataFrame
df_new.show()


root
 |-- _corrupt_record: string (nullable = true)
 |-- petalLength: double (nullable = true)
 |-- petalWidth: double (nullable = true)
 |-- sepalLength: double (nullable = true)
 |-- sepalWidth: double (nullable = true)
 |-- species: string (nullable = true)

+---------------+-----------+----------+-----------+----------+-------+
|_corrupt_record|petalLength|petalWidth|sepalLength|sepalWidth|species|
+---------------+-----------+----------+-----------+----------+-------+
|[              |NULL       |NULL      |NULL       |NULL      |NULL   |
|NULL           |1.4        |0.2       |5.1        |3.5       |setosa |
|NULL           |1.4        |0.2       |4.9        |3.0       |setosa |
|NULL           |1.3        |0.2       |4.7        |3.2       |setosa |
|NULL           |1.5        |0.2       |4.6        |3.1       |setosa |
|NULL           |1.4        |0.2       |5.0        |3.6       |setosa |
|NULL           |1.7        |0.4       |5.4        |3.9       |setosa |
|NULL           |1

In [ ]:
from pyspark.sql.functions import col


# Calculate Pearson Correlation between petalLength and petalWidth
correlation = df_new.stat.corr("petalLength", "petalWidth", method="pearson")
print(f"Pearson Correlation between petalLength and petalWidth: {correlation}")

# Show the columns sepalLength, sepalWidth, and species for rows where petalLength is >= 1.4
filtered_df = df_new.filter(col("petalLength") >= 1.4).select("sepalLength", "sepalWidth", "species")
filtered_df.show()

spark.stop()

Pearson Correlation between petalLength and petalWidth: 0.9628654287041563
+-----------+----------+-------+
|sepalLength|sepalWidth|species|
+-----------+----------+-------+
|        4.6|       3.1| setosa|
|        5.4|       3.9| setosa|
|        5.0|       3.4| setosa|
|        4.9|       3.1| setosa|
|        5.4|       3.7| setosa|
|        4.8|       3.4| setosa|
|        5.7|       4.4| setosa|
|        5.7|       3.8| setosa|
|        5.1|       3.8| setosa|
|        5.4|       3.4| setosa|
|        5.1|       3.7| setosa|
|        5.1|       3.3| setosa|
|        4.8|       3.4| setosa|
|        5.0|       3.0| setosa|
|        5.0|       3.4| setosa|
|        5.2|       3.5| setosa|
|        4.7|       3.2| setosa|
|        4.8|       3.1| setosa|
|        5.4|       3.4| setosa|
|        5.2|       4.1| setosa|
+-----------+----------+-------+
only showing top 20 rows



-->the high value of 0.963 suggests a strong positive correlation between "petalLength" and "petalWidth." As "petalLength" increases, "petalWidth" tends to increase as well.